Blogpost about TimescaleDB integration (#990)
debezium/debezium.github.io: _posts/2024-01-16-Debezium-and-TimescaleDB.adoc

featured: true
author: jpechane
---
In this article, we are going to present and demonstrate a new feature delivered in Debezium 2.4 - the integration with the TimescaleDB database.

link:https://github.com/timescale/timescaledb[TimescaleDB] is an open-source database designed to make SQL scalable for time-series data.
It is implemented as an extension for the PostgreSQL database.
This fact leads us to re-use the standard link:https://debezium.io/documentation/reference/stable/connectors/postgresql.html[Debezium PostgreSQL connector] and implement TimescaleDB support as a link:https://cwiki.apache.org/confluence/display/KAFKA/KIP-66%3A+Single+Message+Transforms+for+Kafka+Connect[single message transform (SMT)].

+++<!-- more -->+++

TimescaleDB provides three basic building blocks/concepts:

* Hypertables
* Continuous aggregates
* Compression

Metadata (catalog) that describes the definitions of the instances and the raw data are typically stored in `_timescaledb_internal_schema`.
link:https://debezium.io/documentation/reference/stable/transformations/timescaledb.html[TimescaleDb SMT] connects to the database and reads and processes the metadata.

> Based on them the raw messages read from the database are enriched with the metadata stored in Kafka Connect headers that create the relation between physical data and the TimescaleDB logical constructs.

Reviewer: ```suggestion
The raw messages read from the database are then enriched with the metadata stored in Kafka Connect headers, creating the relation between the physical data and the TimescaleDB logical constructs.
```

[source, bash]
----
$ docker-compose -f docker-compose-timescaledb.yaml up --build
----

The command will bring up Debezium (Zookeeper, Kafka, Kafka Connect) and the source TimescaleDB database.

The started database is primed with the following database objects:

* Hypertable `conditions` representing temperature and humidity measurements as time-series data; created with DDL `CREATE TABLE conditions (time TIMESTAMPTZ NOT NULL, location TEXT NOT NULL, temperature DOUBLE PRECISION NULL, humidity DOUBLE PRECISION NULL); SELECT create_hypertable('conditions', 'time')`
* A single record of the measurement data (`INSERT INTO conditions VALUES(NOW(), 'Prague', 22.8,  53.3)`)
* PostgreSQL publication used to publish time-series data into replication slot as the demo uses `pgoutput` decoding plugin (`CREATE PUBLICATION dbz_publication FOR ALL TABLES WITH (publish = 'insert, update')`)

In the next step it is necessary to register the Debezium PostgreSQL connector to capture the changes in the database

[source, bash]
----
$ curl -i -X POST -H "Accept:application/json" -H  "Content-Type:application/json" http://localhost:8083/connectors/ -d @register-timescaledb.yaml
----

> The registration request file differs from the regular one with the addition of lines

Reviewer: ```suggestion
The registration request file differs from the regular one with the addition of these lines
```

[source, json]
----
{
    "name": "inventory-connector",
    "config": {
...
        "schema.include.list": "_timescaledb_internal",
        "transforms": "timescaledb",
        "transforms.timescaledb.type": "io.debezium.connector.postgresql.transforms.timescaledb.TimescaleDb",
        "transforms.timescaledb.database.hostname": "timescaledb",
        "transforms.timescaledb.database.port": "5432",
        "transforms.timescaledb.database.user": "postgres",
        "transforms.timescaledb.database.password": "postgres",
        "transforms.timescaledb.database.dbname": "postgres"
    }
}
----

=== Hypertables

> The connector will capture the internal TimescaleDB schema with the physical tables containing the raw data and the `TimescaleDb` SMT will be applied to enrich messages and route them to correctly named topics based on the logical names.

Reviewer: ```suggestion
The connector will capture the internal TimescaleDB schema with the physical tables containing the raw data and the `TimescaleDb` SMT will be applied to enrich messages and route them to the correctly named topics based on the logical names.
```

The SMT configuration options contain information needed to connect to the database.

> In this case, the `conditions` hypertable will be physically stored in `_timescaledb_internal._hyper_1_1_chunk` and when processed by the SMT it will be re-routed to `timescaledb.public.conditions` topic that is named according to fixed configured prefix `timescaledb` and logical name `public.conditions` that conforms to the hypertable name.

Reviewer: ```suggestion
In this case, the `conditions` hypertable will be physically stored in `_timescaledb_internal._hyper_1_1_chunk` and when processed by the SMT, it will be re-routed to `timescaledb.public.conditions` topic that is named according to fixed configured prefix `timescaledb` and logical name `public.conditions` that conforms to the hypertable name.
```

If a new measurement is added and aggregate recomputation is triggered then an updated aggregate is emitted to the topic

[source, bash]
----
postgres=# INSERT INTO conditions VALUES (now(), 'Ostrava', 10, 50);
postgres=# CALL refresh_continuous_aggregate('conditions_summary', CURRENT_DATE, CURRENT_DATE + 1);
----

that looks like

[source, bash]
----
{
   "schema":{
...
   },
   "payload":{
      "before":null,
      "after":{
         "location":"Ostrava",
         "bucket":"2024-01-09T13:00:00.000000Z",
         "avg":10.0,
         "max":10.0,
         "min":10.0
      },
      "source":{
         "version":"2.5.0.Final",
         "connector":"postgresql",
         "name":"dbserver1",
         "ts_ms":1704806938840,
         "snapshot":"false",
         "db":"postgres",
         "sequence":"[\"29727872\",\"29728440\"]",
         "schema":"public",
         "table":"conditions_summary",
         "txId":764,
         "lsn":29728440,
         "xmin":null
      },
      "op":"c",
      "ts_ms":1704806939163,
      "transaction":null
   }
}
----

So the topic contains two or more messages calculated for two different locations.

=== Compression

> The TimescaleDB SMT does not enhance compressed chunks of data (physical table records), only as a by-product of them being stored in hypertable.

Reviewer: ```suggestion
The TimescaleDB SMT does not enhance compressed chunks of data (physical table records), only as a by-product of them being stored in a hypertable.
```

> The compressed data are captured and stored in the Kafka topic.

Reviewer: ```suggestion
The compressed data is captured and stored in the Kafka topic.
```

Typically, messages with compressed chunks are dropped and are not processed by subsequent jobs in the pipeline.

Let's enable compression for the hypertable and compress it
[source, bash]
----
postgres=# ALTER TABLE conditions SET (timescaledb.compress, timescaledb.compress_segment by = 'location');
postgres=# SELECT show_chunks('conditions');
              show_chunks               
----------------------------------------
 _timescaledb_internal._hyper_1_1_chunk
(1 row)

postgres=# SELECT compress_chunk( '_timescaledb_internal._hyper_1_1_chunk');
----

Messages are written to `timescaledb._timescaledb_internal._compressed_hypertable_3`.

Tear down the environment

[source, bash]
----
docker-compose -f docker-compose-timescaledb.yaml down
----

== Conclusion

> In this post, we have demonstrated the capturing of data from TimescaleDB time-series database and their processing by TimescaleDb SMT.

Reviewer: ```suggestion
In this post, we have demonstrated the capturing of data from TimescaleDB time-series database and their processing by the TimescaleDb SMT.
```
